crates/eips/src/eip2718.rs

```diff
@@ -0,0 +1,162 @@
+/// or a receipt with a type flag.
+///
+/// [EIP-2718]: https://eips.ethereum.org/EIPS/eip-2718
+pub trait Decodable2718: Sized {
+    /// Extract the type byte from the buffer, if any. The type byte is the
+    /// first byte, provided that that first byte is 0x7f or lower.
+    fn extract_type_byte(buf: &mut &[u8]) -> Option<u8> {
+        buf.first().copied().filter(|b| *b <= TX_TYPE_BYTE_MAX)
+    }
+
+    /// Decode the appropriate variant, based on the type flag.
+    ///
+    /// This function is invoked by [`Self::decode_2718`] with the type byte, and the tail of the
+    /// buffer.
+    ///
+    /// ## Note
+    ///
+    /// This should be a simple match block that invokes an inner type's RLP decoder.
+    fn typed_decode(ty: u8, buf: &mut &[u8]) -> Result<Self, Eip2718Error>;
+
+    /// Decode the default variant.
+    ///
+    /// This function is invoked by [`Self::decode_2718`] when no type byte can be extracted.
+    fn fallback_decode(buf: &mut &[u8]) -> Result<Self, Eip2718Error>;
+
+    /// Decode an EIP-2718 transaction into a concrete instance
+    fn decode_2718(buf: &mut &[u8]) -> Result<Self, Eip2718Error> {
+        Self::extract_type_byte(buf)
+            .map(|ty| Self::typed_decode(ty, &mut &buf[1..]))
+            .unwrap_or_else(|| Self::fallback_decode(buf))
+    }
+
+    /// Decode an EIP-2718 transaction in the network format.
+    ///
+    /// The network format is the RLP encoded string consisting of the
+    /// type-flag prepneded to an opaque inner encoding. The inner encoding is
+    /// RLP for all current Ethereum transaction types, but may not be in future
+    /// versions of the protocol.
+    fn network_decode(buf: &mut &[u8]) -> Result<Self, Eip2718Error> {
+        let h_decode = &mut *buf;
+        let h = Header::decode(h_decode)?;
+
+        if h.list {
+            return Self::fallback_decode(buf);
+        } else {
+            *buf = h_decode;
+        }
+
+        let pre_len = buf.len();
+        if pre_len == 0 || pre_len < h.payload_length {
+            return Err(alloy_rlp::Error::InputTooShort.into());
+        }
+        let ty = buf[0];
+        let buf = &mut &buf[1..];
+        let tx = Self::typed_decode(ty, buf)?;
+
+        if buf.len() != pre_len - h.payload_length {
+            return Err(alloy_rlp::Error::UnexpectedLength.into());
+        }
+
+        Ok(tx)
+    }
+}
+
+/// Encoding trait for [EIP-2718] envelopes. These envelopes wrap a transaction
+/// or a receipt with a type flag.
+///
+/// [EIP-2718]: https://eips.ethereum.org/EIPS/eip-2718
+pub trait Encodable2718: Sized + Send + Sync + 'static {
+    /// Return the type flag (if any).
+    ///
+    /// This should return `None` for the default (legacy) variant of the
+    /// envelope.
+    fn type_flag(&self) -> Option<u8>;
+
+    /// True if the envelope is the legacy variant.
+    fn is_legacy(&self) -> bool {
+        self.type_flag().is_none()
```

Comment:
shouldn't this also be for Some(0)?

Reply:
I guess so, added in https://github.com/alloy-rs/alloy/commit/1a240d018b77b958290c47fcb4fc2d6d91c2141b